src/main/java/ch/njol/skript/expressions/ExprTarget.java

```diff
@@ -153,21 +165,43 @@ public String toString(@Nullable Event event, boolean debug) {
 	/**
 	 * Gets an entity's target.
 	 *
-	 * @param entity The entity to get the target of
+	 * @param origin The entity to get the target of.
 	 * @param type The exact EntityData to find. Can be null for any entity.
-	 * @return The entity's target
+	 * @return The entity's target.
 	 */
 	@Nullable
 	@SuppressWarnings("unchecked")
-	public static <T extends Entity> T getTarget(LivingEntity entity, @Nullable EntityData<T> type) {
-		if (entity instanceof Mob)
-			return ((Mob) entity).getTarget() == null || type != null && !type.isInstance(((Mob) entity).getTarget()) ? null : (T) ((Mob) entity).getTarget();
-
-		RayTraceResult result = entity.rayTraceEntities(SkriptConfig.maxTargetBlockDistance.value());
+	public static <T extends Entity> T getTarget(LivingEntity origin, @Nullable EntityData<T> type) {
+		if (origin instanceof Mob)
+			return ((Mob) origin).getTarget() == null || type != null && !type.isInstance(((Mob) origin).getTarget()) ? null : (T) ((Mob) origin).getTarget();
+		Location location = origin.getLocation();
+		RayTraceResult result = null;
+		double raySize = 0.0D;
+		// TODO when DisplayData is added.
+//		if (type.getClass().equals(DisplayData.class))
+//			raySize = 1.0D;
+		Predicate<Entity> predicate = entity -> {
+			if (entity.equals(origin))
+				return false;
+			if (type != null && !type.isInstance(entity))
+				return false;
+			if (entity instanceof Player && ((Player) entity).getGameMode() == GameMode.SPECTATOR)
+				return false;
+			return true;
+		};
+		if (!ignoreBlocks) {
+			RayTraceResult blockResult = origin.getWorld().rayTraceBlocks(origin.getEyeLocation(), location.getDirection(), targetBlockDistance);
+			Block hitBlock = blockResult.getHitBlock();
+			if (hitBlock != null) {
+				result = origin.getWorld().rayTraceEntities(origin.getEyeLocation(), location.getDirection(), location.distance(hitBlock.getLocation()), raySize, predicate);
```

Comment:
The max distance should use the precise hit position on the block, not the hit block's location. Using the block's location results in the entity trace max distance being up to 0.85-ish meters different from the actual position on the block the player is looking.